ComponentKit/Utilities/CKComponentAction.mm

```diff
@@ -125,7 +131,38 @@ - (instancetype)initWithAction:(CKComponentAction)action
 - (void)handleControlEventFromSender:(UIControl *)sender withEvent:(UIEvent *)event
 {
   // If the action can be handled by the sender itself, send it there instead of looking up the chain.
-  CKComponentActionSend(_action, sender.ck_component, event, CKComponentActionSendBehaviorStartAtSender);
+  _action.send(sender.ck_component, CKComponentActionSendBehaviorStartAtSender, event);
+}
+
+#pragma mark - Debug Helpers
+
+void _CKTypedComponentDebugCheckTargetSelector(id target, SEL selector, const std::vector<const char *> &typeEncodings)
+{
+  // In DEBUG mode, we want to do the minimum of type-checking for the action that's possible in Objective-C. We
+  // can't do exact type checking, but we can ensure that you're passing the right type of primitives to the right
+  // argument indices.
+  CKCAssert([target respondsToSelector:selector], @"Target does not respond to selector for component action. -[%@ %@]", [target class], NSStringFromSelector(selector));
+
+  NSMethodSignature *signature = [target methodSignatureForSelector:selector];
+
+  CKCAssert(typeEncodings.size() + 3 >= signature.numberOfArguments, @"Expected action method %@ to take less than %lu arguments, but it suppoorts %lu", NSStringFromSelector(selector), typeEncodings.size(), (unsigned long)signature.numberOfArguments - 3);
+
+  CKCAssert(signature.methodReturnLength == 0, @"Component action methods should not have any return value. Any objects returned from this method will be leaked.");
+
+  for (int i = 0; i + 3 < signature.numberOfArguments; i++) {
+    const char *methodEncoding = [signature getArgumentTypeAtIndex:i + 3];
+    const char *typeEncoding = typeEncodings[i];
+
+    CKCAssert(strcmp(methodEncoding, typeEncoding) == 0, @"Implementation of %@ does not match expected types.\nExpected type %s, got %s", NSStringFromSelector(selector), typeEncoding, methodEncoding);
+  }
```

Comment:
This method provides really nice warnings to you if you're going to do something stupid, like pass a method with multiple arguments to an action that only takes 1. We decided to make requirements a bit weak on objc methods. For instance, if your action is defined:

```
CKTypedComponentAction<id, id, id>
```

You could pass in a method that took only one:

```
[SomeComponent
newWithAction:{target, @selector(sender:context1:)}]
```

This will invoke the selector with the sender, and the first param present. This is built to mirror the behavior of performSelector:withObject: which we used before this diff.